src/client/java/dev/creesch/storage/ChatMessageRepository.java

```diff
@@ -0,0 +1,195 @@
+package dev.creesch.storage;
+
+import com.google.gson.JsonObject;
+import dev.creesch.model.ChatMessagePayload;
+import dev.creesch.model.WebsocketJsonMessage;
+import dev.creesch.util.NamedLogger;
+import net.fabricmc.loader.api.FabricLoader;
+import org.sqlite.SQLiteDataSource;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static dev.creesch.model.WebsocketMessageBuilder.createHistoricChatMessage;
+
+public class ChatMessageRepository {
+    private final SQLiteDataSource dataSource;
+    private static final String DB_NAME = "chat_messages.db";
+    private static final String DATA_DIR = "web-chat";
+
+    private static final int CURRENT_SCHEMA_VERSION = 1;
+
+    private static final NamedLogger LOGGER = new NamedLogger("web-chat");
+
+    public ChatMessageRepository() {
+        Path databasePath = FabricLoader.getInstance()
+            .getGameDir()
+            .resolve(DATA_DIR)
+            .resolve(DB_NAME);
+
+        try {
+            Files.createDirectories(databasePath.getParent());
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to create data for web-chat database directory", e);
+        }
+
+
+        dataSource = new SQLiteDataSource();
+        dataSource.setUrl("jdbc:sqlite:" + databasePath);
+        initializeDatabase();
+    }
+
+    private void initializeDatabase() {
+        try (Connection conn = dataSource.getConnection()) {
+            conn.createStatement().execute("""
+                    CREATE TABLE IF NOT EXISTS messages (
+                        id INTEGER PRIMARY KEY AUTOINCREMENT,
+                        timestamp BIGINT NOT NULL,
+                        server_id TEXT NOT NULL,
+                        server_name TEXT NOT NULL,
+                        message_id TEXT NOT NULL,
+                        message_json TEXT NOT NULL,
+                        minecraft_version TEXT
+                    )
+                """);
+
+            // Create composite index for server_id + timestamp queries
+            conn.createStatement().execute(
+                "CREATE INDEX IF NOT EXISTS idx_server_id_timestamp ON messages(server_id, timestamp DESC)"
+            );
+
+            // Version table
+            conn.createStatement().execute("""
+                    CREATE TABLE IF NOT EXISTS schema_version (
+                        version INTEGER PRIMARY KEY
+                    )
+                """);
```

Comment:
This `"""` spacing seems strange to me but I won't fight it.

Reply:
Tbh, it is new to me, I have mostly worked with java 11 recently where this isn't a thing yet.